crates/signature-mpc/src/twopc_mpc_protocols/dwallet_transfer.rs

```diff
@@ -0,0 +1,361 @@
+use enhanced_maurer::{
+    encryption_of_discrete_log, EnhanceableLanguage, EnhancedLanguage, Proof,
+    PublicParameters as MaurerPublicParameters, WitnessSpaceGroupElement,
+};
+use group::{secp256k1, GroupElement, Samplable};
+use homomorphic_encryption::{
+    AdditivelyHomomorphicEncryptionKey,
+    GroupsPublicParametersAccessors as PublicParametersAccessors,
+};
+use maurer::{language, SOUND_PROOFS_REPETITIONS};
+use proof::range;
+use proof::range::bulletproofs;
+use proof::range::bulletproofs::{COMMITMENT_SCHEME_MESSAGE_SPACE_SCALAR_LIMBS, RANGE_CLAIM_BITS};
+use rand_core::OsRng;
+use serde::{Deserialize, Serialize};
+use std::marker::PhantomData;
+use tiresias::{
+    CiphertextSpaceGroupElement, CiphertextSpaceValue, DecryptionKey, EncryptionKey,
+    LargeBiPrimeSizedNumber, PlaintextSpaceGroupElement, RandomnessSpaceGroupElement,
+};
+use twopc_mpc::paillier::PLAINTEXT_SPACE_SCALAR_LIMBS;
+use twopc_mpc::secp256k1::paillier::bulletproofs::ProtocolPublicParameters;
+pub use twopc_mpc::secp256k1::{Scalar, SCALAR_LIMBS};
+
+pub type LangPublicParams = language::PublicParameters<SOUND_PROOFS_REPETITIONS, Lang>;
+
+pub const RANGE_CLAIMS_PER_SCALAR: usize =
+    Uint::<{ secp256k1::SCALAR_LIMBS }>::BITS / RANGE_CLAIM_BITS;
+
+pub type Lang = encryption_of_discrete_log::Language<
+    { tiresias::PLAINTEXT_SPACE_SCALAR_LIMBS },
+    { U256::LIMBS },
+    secp256k1::GroupElement,
+    EncryptionKey,
+>;
+
+pub type SecretShareProof = Proof<
+    { maurer::SOUND_PROOFS_REPETITIONS },
+    RANGE_CLAIMS_PER_SCALAR,
+    COMMITMENT_SCHEME_MESSAGE_SPACE_SCALAR_LIMBS,
+    bulletproofs::RangeProof,
+    RandomnessSpaceGroupElement,
+    Lang,
+    PhantomData<()>,
+>;
+pub(crate) type EnhancedLang<
+    const REPETITIONS: usize,
+    const NUM_RANGE_CLAIMS: usize,
+    UnboundedWitnessSpaceGroupElement,
+    Lang,
+> = EnhancedLanguage<
+    REPETITIONS,
+    NUM_RANGE_CLAIMS,
+    { COMMITMENT_SCHEME_MESSAGE_SPACE_SCALAR_LIMBS },
+    range::bulletproofs::RangeProof,
+    UnboundedWitnessSpaceGroupElement,
+    Lang,
+>;
+
+pub struct ProofPublicOutput {
+    pub proof: SecretShareProof,
+    pub encrypted_discrete_log: CiphertextSpaceValue,
+    pub range_proof_commitment: range::CommitmentSchemeCommitmentSpaceValue<
+        { COMMITMENT_SCHEME_MESSAGE_SPACE_SCALAR_LIMBS },
+        { RANGE_CLAIMS_PER_SCALAR },
+        bulletproofs::RangeProof,
+    >,
+}
+
+fn pad_vector(vec: Vec<u8>) -> Vec<u8> {
+    let target_length = 256;
+    if vec.len() >= target_length {
+        return vec;
+    }
+    let mut padded_vec = vec![0; target_length - vec.len()];
+    padded_vec.extend(vec);
+    padded_vec
+}
+
+pub fn generate_keypair() -> (Vec<u8>, Vec<u8>) {
+    let (public_key, private_key) = DecryptionKey::generate(&mut OsRng).unwrap();
+    let private_key = bcs::to_bytes(&private_key.secret_key).unwrap();
+    let public_key = bcs::to_bytes(&public_key).unwrap();
+    (public_key, private_key)
```

Comment:
This is coupled to paillier/tiresias, maybe we should use generic traits instead

Reply:
Actually it seems I didn't add a `generate` API there, so for now we can keep it as-is